### ISSUE TITLE
Make it ready for v0.9.0

### ISSUE THREAD

**Issue by koenichiwa** [zslayton/cron#82]
Because I fixed two of the metaissues in #67 and I would like to use them easily in my own personal project. I was wondering what could be done to make this project ready for v0.9.0

I was thinking a couple of things:

- [ ] Run a more heftier version of clippy against the codebase.
     -  I tried running clippy::pedantic and some parts of clippy::nursery, and even though it did have a couple of false positives it did help me spot a couple of issues. E.g. unneeded borrows, changing `.map(f).unwrap_or_else(g)` into `.map_or_else(g,f)`, or with ranges: `(min..max + 1)` into `(min..=max)`
- [ ] Small refactor
     - Giving Field it's own file and ScheduleFields it's own file. Also changing the location of things like `FromString for Schedule` to the schedule.rs file.

If there are any other issues that need to be addressed I would love to see it.

Edit:
- [x] Relicense to dual MIT/Apache-2

**Comment by zslayton** [comment]
These are both good changes to make. One other I'd like to squeeze into v0.9.0 is a fix for #8, changing the license from MIT to dual MIT/Apache-2. The changes you've made since v0.8.0 are substantial though, so I'd be ok doing a release without that change if you'd prefer.

**Comment by koenichiwa** [comment]
Yeah sure! I'm not really at home with licensing though, so if you can do that, then I can work on the other two issues?

**Comment by zslayton** [comment]
> I'm not really at home with licensing though

Me neither. :) I can take care of it, sure.

**Comment by koenichiwa** [comment]
Quick question about `OrdinalIter` `OrdinalRangeIter` etc. Shouldn't they actually reside in the ordinal module?

**Comment by zslayton** [comment]
> Quick question about OrdinalIter OrdinalRangeIter etc. Shouldn't they actually reside in the ordinal module?

Sounds sensible to me.

> Maybe this is why my implementation of the queries was so slow? Can it be because the current implementation of iter copies every Ordinal before returning it?

It's impossible to rule it out without testing/measurement, but that would surprise me. An `Ordinal` is a `u32`, so it fits in a register on x86_64 machines. Copying it should be trivial.

> Isn't this exactly a reason why this should be implemented as a type definition like above, so the end user can chose whether to copy or not?

The copying that's happening is basically a dereference of the `&u32` provided by the `BTreeSet` iterator. If we returned the `&u32` directly to the user, they'd just end up doing the dereference themselves.

I'm happy to be proven wrong if you'd like to dig into it!

**Comment by koenichiwa** [comment]
Welp, I guess your right.
I have a counter offer though, what if it gets implemented as
```rust
pub type OrdinalIter<'a> = Cloned<btree_set::Iter<'a, Ordinal>>;
pub type OrdinalRangeIter<'a> = Cloned<btree_set::Range<'a, Ordinal>>;
```
That way you get all `btree_set::Iter`s traits for free

**Comment by koenichiwa** [comment]
Nitpick, but you forgot the license change in Cargo.toml

**Comment by zslayton** [comment]
> ```
> pub type OrdinalIter<'a> = Cloned<btree_set::Iter<'a, Ordinal>>;
> pub type OrdinalRangeIter<'a> = Cloned<btree_set::Range<'a, Ordinal>>;
> ```

Based on a hazy recollection of implementing this years ago, I believe my main concerns with using a type alias were that:
* It leaks implementation details to users of the crate. At least at the time, error messages would contain the aliased type names (`Cloned<btree_set::Iter<'a, Ordinal>>`), making it clear how internals were being implemented and effectively making those types part of the public API.
* It prevented me from having control over what `OrdinalIter` could and couldn't do. If `btree_set::Iter` added or removed functionality, `cron`'s API would change whether I liked it or not.

Combined with the fact that the compiler will reduce the layout of a struct with a single field into that field's layout, it seemed reasonable to just make my own wrapper type.

I wish Rust offered [a version of `newtype`](https://rust-unofficial.github.io/patterns/patterns/newtype.html) that was syntactically closer to a type alias. 😞 

> as a bonus you get all btree_set::Iters traits for free whithout anything changing

That's compelling! Unfortunately, I think the public API exposure is enough of a problem that it's worth exposing those individually/manually. Which traits have you found yourself missing?

**Comment by koenichiwa** [comment]
> It leaks implementation details to users of the crate. At least at the time, error messages would contain the aliased type names (Cloned<btree_set::Iter<'a, Ordinal>>), making it clear how internals were being implemented and effectively making those types part of the public API.

I honestly didn't know that. But I get that that is a concern.

> It prevented me from having control over what OrdinalIter could and couldn't do. If btree_set::Iter added or removed functionality, cron's API would change whether I liked it or not.

I doubt that those kind of breaking changes to the standard library would happen without an opt-in. Maybe if you change the rust edition. Otherwise a whole boatload of rust libraries would have the same problem. One of the first lines of the std::collections module README states: 
>By using the standard implementations, it should be possible for two libraries to communicate without significant data conversion.

So I thought to live by that "credo" as much as possible

> That's compelling! Unfortunately, I think the public API exposure is enough of a problem that it's worth exposing those individually/manually. Which traits have you found yourself missing?

I'm not missing anything in particular. I was just trying to make everything as neat as possible, so that you reason about the project a bit more easily. I just thought that the implementation of a range and an iter was outside of the scope of the project.
If I'm going to refactor it without a typealias, then I just need to either make the field pub, or create a pub constructor. The module is private though, so it will only be exposed to the crate.

**Comment by koenichiwa** [comment]
On an unrelated note, I have two nits.
-  `OrdinalRangeIter` does not implement `Iter`, shouldn't it get renamed to `OrdinalRange`
-  `ScheduleIterator` is returned by some public functions of `Schedule`. But because the type itself is not public, the end user can't read up on it in the docs. Shouldn't it be public (with a private constructor of course)

Sorry if coming of as too critical about a project that isn't mine, I'm just trying to help.

**Comment by zslayton** [comment]
It's likely to be a day or two before I have a chance to go through the changes you're describing.

> Sorry if coming of as too direct or too critical about a project that isn't mine, I'm just trying to help.

Not at all, I appreciate it. It's great to have a second opinion on these things, especially if the library ever hopes to release a `v1.0`!

**Comment by zslayton** [comment]
> If I'm going to refactor it without a typealias, then I just need to either make the field pub, or create a pub constructor.

I think I've missed a beat: what's the use case for users creating their own ordinal iterators?

> OrdinalRangeIter does not implement Iter, shouldn't it get renamed to OrdinalRange

I'm not aware of a trait called `Iter`. If you're referring to `Iterator`, though, it does implement that:

https://github.com/zslayton/cron/blob/2bbb4f224f8b21beb4b3c6c04381c5e93cfd91a0/src/time_unit/mod.rs#L42-L51

(Sorry if there's actually an `Iter` trait--could you point me to it?)

> ScheduleIterator is returned by some public functions of Schedule. But because the type itself is not public, the end user can't read up on it in the docs. Shouldn't it be public (with a private constructor of course), just so users know what they're working with?

Huh! While the type itself is public:

https://github.com/zslayton/cron/blob/d0049fc4150b1e1e15408b4d118df7100584b328/src/schedule.rs#L342-L349

(which allows it to be returned by public `fn`s in `Schedule`), the `schedule` module it lives in is not, which prevents it from appearing in documentation. Interesting. I think the simplest fix is to re-export `ScheduleIterator` in `lib.rs` like we do for `Schedule` and `TimeUnitSpec`:

https://github.com/zslayton/cron/blob/d0049fc4150b1e1e15408b4d118df7100584b328/src/lib.rs#L50-L51

Good eye!

**Comment by koenichiwa** [comment]
> think I've missed a beat: what's the use case for users creating their own ordinal iterators?

Oh no, I meant exactly the opposite! If I refactor it out to the ordinal mod, it needs a private function so it can be used in the time_unit mod. But it should stay private for the user.

> I'm not aware of a trait called Iter. If you're referring to Iterator, though, it does implement that:

Oops, it seems I've made a mistake. I was lookming at the Range and Iter **structs** in BTreeSet. As far as I can see, range isn't even an existing trait. But it still might be an idea to follow convention and call it OrdinalRange.


Concerning the re-exports, the same goes for Ordinal, OrdinalIter and OrdinalRangeIter

**Comment by zslayton** [comment]
Sorry, I let the ball drop on this thread. 😞 I'll try to revisit it later this week.

For the moment: I've [released v0.9.0](https://github.com/zslayton/cron/releases/tag/v0.9.0) to make the fix from #84 available. I'll do another release once this change set lands. You dominated the release notes for this one, thanks again for all your help!

**Comment by koenichiwa** [comment]
Nothing to worry about. My study picked up the pace again, so I'm focusing on other stuff currently. We'll get back to this later 🙂

**Comment by koenichiwa** [comment]
Having said that. I was bored, and thought I would implement ordinal.rs back to the struct/impl implementation. I left the type aliases commented though, so you can make a comparison, tell me which direction we should take, and what should be deleted
